File: Problems/No.2099_Find_Subsequence_of_Length_K_With_the_Largest_Sum/find_subsequence_of_length_k_with_the_largest_sum.cpp
class Solution {
public:
    vector<int> maxSubsequence(vector<int>& nums, int k) {
        
        // parameters
        std::vector<int> MS_Sequence;
        std::priority_queue<int> MAX_Heap;
        std::map<int, int> LargestSum;

        // start computing
        // create MAX-Heap
        for (auto &num: nums) MAX_Heap.push(num);

        // find largest sum
        for (int i=0; i<k; i++) {

            LargestSum[MAX_Heap.top()]++;
            MAX_Heap.pop();

        }

        // push in original order
        for (auto &num: nums) {
            if (LargestSum[num] > 0) {

                MS_Sequence.push_back(num);
                LargestSum[num]--;

            }
        }

        return MS_Sequence;

    }
};

File: Problems/No.2099_Find_Subsequence_of_Length_K_With_the_Largest_Sum/find_subsequence_of_length_k_with_the_largest_sum.py
class Solution(object):
    def maxSubsequence(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """

        # parameters
        MS_Sequence = []
        MAX_Heap = deepcopy(nums)
        LargestSum = {}
        
        # start computing
        # create MAX-Heap
        for i in range(len(MAX_Heap)):
            MAX_Heap[i] *= -1
        heapq.heapify(MAX_Heap)

        # find largest sum
        k_largest = heapq.nsmallest(k, MAX_Heap)
        for num in k_largest:
            num *= -1
            LargestSum[num] = LargestSum.get(num, 0) + 1

        # push in original order
        for num in nums:
            if (LargestSum.get(num, 0) > 0):
                MS_Sequence.append(num)
                LargestSum[num] -= 1

        return MS_Sequence
